harvester_e2e_tests/integration/test_vm_snapshot.py

```diff
@@ -0,0 +1,517 @@
+            code, data = api_client.vm_snapshots.get(vm_snapshot_name)
+            if data.get("status", {}).get("readyToUse"):
+                break
+            print(f"waiting for {vm_snapshot_name} to be ready")
+            sleep(3)
+        else:
+            raise AssertionError(f"timed out waiting for {vm_snapshot_name} to be ready")
+
+        assert 200 == code
+        assert data.get("status", {}).get("readyToUse") is True
+
+    @pytest.mark.dependency(depends=["source_vm_snapshot"])
+    def test_restore_into_new_vm_from_vm_snapshot(self, api_client,
+                                                  restored_from_snapshot_vm, source_vm,
+                                                  ssh_keypair, host_shell,
+                                                  vm_shell, wait_timeout):
+        """
+        Test that restoring the `vm-snapshot` into a new virtual
+        machine results in a virtual machine with the expected
+        well-known file (`test.txt`) with the expected contents
+        (`123`).
+
+        Prerequisites:
+        1. The source VM from the first test case and its
+           snapshot (`vm-snapshot`).
+        2. A VM restored from `vm-snapshot`.
+        """
+
+        _, ssh_user = source_vm
+
+        def actassert(sh):
+            out, _ = sh.exec_command("cat test.txt")
+            assert "123" in out
+
+        vm_shell_do(restored_from_snapshot_vm, api_client,
+                    host_shell, vm_shell,
+                    ssh_user, ssh_keypair,
+                    actassert, wait_timeout)
+
+    @pytest.mark.dependency(name="replaced_source_vm", depends=["source_vm_snapshot"])
+    def test_replace_vm_with_vm_snapshot(self, api_client,
+                                         source_vm, vm_snapshot_name,
+                                         ssh_keypair, host_shell,
+                                         vm_shell, wait_timeout):
+        name, ssh_user = source_vm
+
+        stop_vm(name, api_client, wait_timeout)
+
+        """
+        Test that the Harvester API rejects a `replace`
+        VirtualMachineRestore where the deletePolicy is
+        not `retain`.
+
+        Prequisites:
+        1. The original VM (`source-vm`) and snapshot (`vm-snapshot`)
+        from the first test case.
+        """
+        spec = api_client.backups.RestoreSpec.for_existing(delete_volumes=True)
+        code, data = api_client.vm_snapshots.restore(vm_snapshot_name, spec)
+
+        reason = data.get("message")
+
+        wantmsg = "Delete policy with backup type snapshot"
+        " for replacing VM is not supported"
+
+        assert wantmsg in reason
+        assert 422 == code
+
+        """
+        Test that the original virtual machine can be replaced
+        from its original snapshot (`vm-snapshot`) and that
+        the snapshot's data contains the well-known file (`test.txt`)
+        and its expected contents (`123`).
+
+        Prerequisites:
```

Comment:
same as previous `restore to new`, we didn't modify the VM after snapshot, so we can't guarantee the VM really restored from the snapshot or not.